### ISSUE TITLE
js-output adding a deadly line of pure nonsens

### ISSUE THREAD

**Issue by dy-soft** [microsoft/TypeScript#26046]
Script:
An Electron Renderer-Process-Script, neither containing nor intended to contain any exports. 

Issue:
Has not showed up since about one day or so.

.ts source:
0        import { ipcRenderer } from "electron";
1…     «« all the following code is working fine »» 

.js tsc output:
0              "use strict";
added    // the next line is producing an error, an Object named 'export' does not exist
1               Object.defineProperty(exports, "__esModule", { value: true });
2              var electron_1 = require("electron");

Debugging:
It can be run when that .js-line 1 has not been forgotten to remove after each change of the source…

Version 3.1.0-dev.20180728

defineProperty(exports, "__esModule"

**Code**
.ts
see above, no.zero[zero]=>{ .ts statement } is producing a nonsensical 'export' .js-output-line 

**Expected behavior:**

not to produce such nonsense

**Actual behavior:**

producing such nonsense

**Playground Link:** 

**Related Issues:**

**Comment by dy-soft** [comment]
I forgot the tsconfig.json:
{
  "compilerOptions": {
    /* Basic Options */
    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
    // "lib": [],                             /* Specify library files to be included in the compilation. */
    // "allowJs": true,                       /* Allow javascript files to be compiled. */
    // "checkJs": true,                       /* Report errors in .js files. */
    // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
    "declaration": true,                      /* Generates corresponding '.d.ts' file. */
    // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
    "sourceMap": true,                        /* Generates corresponding '.map' file. */
    // "outFile": "./",                       /* Concatenate and emit output to single file. */
    // "outDir": "./",                        /* Redirect output structure to the directory. */
    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
    // "composite": true,                     /* Enable project compilation */
    // "removeComments": true,                /* Do not emit comments to output. */
    // "noEmit": true,                        /* Do not emit outputs. */
    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */

    /* Strict Type-Checking Options */
    // "strict": true,                        /* Enable all strict type-checking options. */
    // "noImplicitAny": true,                 /* Raise error on expressions and declarations with an implied 'any' type. */
    // "strictNullChecks": true,              /* Enable strict null checks. */
    // "strictFunctionTypes": true,              /* Enable strict checking of function types. */
    // "strictPropertyInitialization": true,     /* Enable strict checking of property initialization in classes. */
    // "noImplicitThis": true,                   /* Raise error on 'this' expressions with an implied 'any' type. */
    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */

    /* Additional Checks */
    // "noUnusedLocals": true,                /* Report errors on unused locals. */
    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */

    /* Module Resolution Options */
    // "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
    // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
    // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
    // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
    // "typeRoots": [],                       /* List of folders to include type definitions from. */
    // "types": [],                           /* Type declaration files to be included in compilation. */
    // "allowSyntheticDefaultImports": true,  /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
    // "esModuleInterop": true,               /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
    // "preserveSymlinks": true,              /* Do not resolve the real path of symlinks. */

    /* Source Map Options */
    // "sourceRoot": "./",                    /* Specify the location where debugger should locate TypeScript files instead of source locations. */
    // "mapRoot": "./",                       /* Specify the location where debugger should locate map files instead of generated locations. */
    // "inlineSourceMap": true,               /* Emit a single file with source maps instead of having a separate file. */
    // "inlineSources": true,                 /* Emit the source alongside the sourcemaps within a single file; requires '--inlineSourceMap' or '--sourceMap' to be set. */

    /* Experimental Options */
    // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
    // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
  },
  "include": [
    "../BHSA_2017" // current project folder
    //"../**/*.ts"
  ]
}

// thanks and blessings

**Comment by DanielRosenwasser** [comment]
What environment are you running in that `exports` isn't defined?

**Comment by kitsonk** [comment]
I would assume Electron on the render side based on:

> An Electron Renderer-Process-Script

Which I suspect is due to the OP not loading the module in render via `require()` and likely trying it via a `<script>` tag.

For Electron, you _might_ want to consider the module format `umd` to avoid this problem, but if it is a "main" type module, you will still have to invoke it somehow on the render side.

Either way, this is a support question really, not a TypeScript bug.

**Comment by dy-soft** [comment]
Thanks for your interest you two, and sorry for the delay.

I'm owning a Mac-Pro 5, even if aging, still very reliable, having all updates, including the nightly build. I have uninstalled and reinstalled the entire NPM cacke several times, global this and local that, local this and global that, including Node 10, is bringing sometimes disadvantages, but never advantages.

Thanks for your tip, kitsonk, I will ty and let you know.

The process:
Electron is somehow running through a native app, and this is bringing some restrictions, at least on my good old boy. 

A. Setting a breakpoint during the initialization, is without any effect.  Setting therein  a "debugger" statement, makes them working, and after this vacation, the breakpoints are working then too.

B. Trying to use another process, by packing the entire script into a "window.setTimeout(function(){««Box»»}, «1-1000»)" is not changing anything. 

C. If there is a mistake in the initialization, the error won't be displayed in VSC, but only in the Dev-Tools, stealing half of the display.

D.  A humanitarian solution is to display an extra sign, indicating if the initialization has succeeded or not. This is really important, when after each and every modification, the neither needed nor working "exports-line" in the js-output has to be commented out, or could be forgotten. 

As you said right, and I've not figured out by myself, it is more an Electron, and less a TS problem, I will report it there as well. But the questions are still raising, about a fail-save philosophy... Why on heaven and earth to add such an "export line", if none is needed, and that process is happy to pass this in its idle-thrust. As much as I'm understanding, compilers are first reading the source, then based on this, doing the next actions, mixed with the options. If there is nothing about "export" around, the corresponding variables should be set in such a manner, and be remembered until the output. I think to add what is needed, means as well, not to add what is not needed, at least for myself. And I don't think that the *.ts and its *.js-birth are a very far or complicated way apart from another.

Thanks anyway to both of you, and never forget to laugh.

**Comment by dy-soft** [comment]
I'm new on GitHub, and not sure if you have been notified from there, where
I wrote the following:

With greetings from Switzerland

Thanks for your interest you two, and sorry for the delay.

I'm owning a Mac-Pro 5, even if aging, still very reliable, having all
updates, including the nightly build. I have uninstalled and reinstalled
the entire NPM cacke several times, global this and local that, local this
and global that, including Node 10, is bringing sometimes disadvantages,
but never advantages.

Thanks for your tip, kitsonk, I will ty and let you know.

The process:
Electron is somehow running through a native app, and this is bringing some
restrictions, at least on my good old boy.

A. Setting a breakpoint during the initialization, is without any effect.
Setting therein a "debugger" statement, makes them working, and after this
vacation, the breakpoints are working then too.

B. Trying to use another process, by packing the entire script into a
"window.setTimeout(function(){««Box»»}, «1-1000»)" is not changing anything.

C. If there is a mistake in the initialization, the error won't be
displayed in VSC, but only in the Dev-Tools, stealing half of the display.

D. A humanitarian solution is to display an extra sign, indicating if the
initialization has succeeded or not. This is really important, when after
each and every modification, the neither needed nor working "exports-line"
in the js-output has to be commented out, or could be forgotten.

As you said right, and I've not figured out by myself, it is more an
Electron, and less a TS problem, I will report it there as well. But the
questions are still raising, about a fail-save philosophy... Why on heaven
and earth to add such an "export line", if none is needed, and that process
is happy to pass this in its idle-thrust. As much as I'm understanding,
compilers are first reading the source, then based on this, doing the next
actions, mixed with the options. If there is nothing about "export" around,
the corresponding variables should be set in such a manner, and be
remembered until the output. I think to add what is needed, means as well,
not to add what is not needed, at least for myself. And I don't think that
the *.ts and its *.js-birth are a very far or complicated way apart from
another.

Thanks anyway to both of you, and never forget to laugh.

Am Mo., 30. Juli 2018 um 05:04 Uhr schrieb Kitson Kelly <
notifications@github.com>:
